jvm and webserver validation for service name conflict

rahulsayini opened the pull request.

rahulsayini committed:
jvm and webserver validation for service name conflict

jeddcuison commented on jwala-services/src/main/java/com/cerner/jwala/service/configuration/service/AemServiceConfiguration.java:
```diff
@@ -97,23 +97,23 @@
 @EnableAsync
 @EnableScheduling
 @ComponentScan({"com.cerner.jwala.service.webserver.component",
-                "com.cerner.jwala.service.state",
```
The alignment was fine already, I don't think you needed to change it.

jeddcuison commented on jwala-services/src/main/java/com/cerner/jwala/service/configuration/service/AemServiceConfiguration.java:
```diff
@@ -170,12 +170,12 @@ public JvmService getJvmService(final GroupPersistenceService groupPersistenceSe
                 messagingTemplate, groupStateNotificationService, resourceService,
                 clientFactoryHelper, topicServerStates, jvmControlService, binaryDistributionService, binaryDistributionLockManager,
-                historyFacadeService, fileUtility);
+                historyFacadeService, webServerPersistenceService,fileUtility);
```
Try to be consistent with spaces, add a space after webServerPersistenceService

jeddcuison commented on jwala-services/src/main/java/com/cerner/jwala/service/configuration/service/AemServiceConfiguration.java:
```diff
@@ -191,21 +191,23 @@ public BalancerManagerService getBalancermanagerService(final GroupService group
                                                 final InMemoryStateManagerService<Identifier<WebServer>, WebServerReachableState> inMemoryStateManagerService,
-                                                @Value("${paths.resource-templates:../data/templates}") final String templatePath) {
+                                                @Value("${paths.resource-templates:../data/templates}") final String templatePath,
+                                                final JvmPersistenceService jvmPersistenceService) {
```
Although I prefer constructor injection, I remember the team decided to do class variable auto-wiring instead to keep the constructor from having too many parameters (which is a good point) 

    class WebServerServiceImpl 
       
        @Autowired
        private JvmPersistenceService jvmPersistenceService;

        ....

    }

However, please confirm with the others if you can do constructor injection for now since we're in a tight deadline

jeddcuison commented on jwala-services/src/main/java/com/cerner/jwala/service/configuration/service/AemServiceConfiguration.java:
```diff
@@ -232,8 +234,8 @@ public ApplicationService getApplicationService(final JvmPersistenceService jvmP
     public JvmControlService getJvmControlService(
-                                                  final JvmStateService jvmStateService,
-                                                  final HistoryFacadeService historyFacadeService) {
+            final JvmStateService jvmStateService,
```
Since you modified the formatting already, wouldn't it be better if you formatted it this way ?

    public JvmControlService getJvmControlService(final JvmStateService jvmStateService, 
                                                  final HistoryFacadeService historyFacadeService) {
        ....
    }

jeddcuison commented on jwala-services/src/main/java/com/cerner/jwala/service/jvm/impl/JvmServiceImpl.java:
```diff
@@ -115,11 +120,18 @@ public JvmServiceImpl(final JvmPersistenceService jvmPersistenceService,
     protected Jvm createJvm(final CreateJvmRequest aCreateJvmRequest) {
+        List<WebServer> webServerList = webServerPersistenceService.getWebServers();
+        for (WebServer webserver : webServerList) {
+            if (aCreateJvmRequest.getJvmName().equals(webserver.getName())) {
```
Use equalsIgnoreCase since we don't want the names to be case sensitive.

jeddcuison commented on jwala-services/src/main/java/com/cerner/jwala/service/jvm/impl/JvmServiceImpl.java:
```diff
@@ -115,11 +120,18 @@ public JvmServiceImpl(final JvmPersistenceService jvmPersistenceService,
     }
 
 
     protected Jvm createJvm(final CreateJvmRequest aCreateJvmRequest) {
```
I can't help notice that we have a createJvm method in this service. The purpose of the JvmStateService is to provide state related functionality. Can you ask @akinny why it's here ? I would prefer it to be in JvmService.

jeddcuison commented on jwala-services/src/main/java/com/cerner/jwala/service/jvm/impl/JvmServiceImpl.java:
```diff
@@ -115,11 +120,18 @@ public JvmServiceImpl(final JvmPersistenceService jvmPersistenceService,
+        List<WebServer> webServerList = webServerPersistenceService.getWebServers();
+        for (WebServer webserver : webServerList) {
+            if (aCreateJvmRequest.getJvmName().equals(webserver.getName())) {
+                throw new EntityExistsException("Webserver already exists with this name "+ aCreateJvmRequest.getJvmName());
```
In addition I'm not in favor of using EntityExistException because at first glance of an exception, it doesn't tell where the exception occurred. I would prefer that you use JvmStateServiceException.

jeddcuison commented on jwala-services/src/main/java/com/cerner/jwala/service/jvm/impl/JvmServiceImpl.java:
```diff
@@ -143,7 +155,6 @@ protected Jvm createAndAssignJvm(final CreateJvmAndAddToGroupsRequest aCreateAnd
     public Jvm createJvm(CreateJvmAndAddToGroupsRequest createJvmAndAddToGroupsRequest, User user) {
```
Again ask @akinny if this can be transferred to JvmService.

jeddcuison commented on jwala-services/src/main/java/com/cerner/jwala/service/jvm/impl/JvmServiceImpl.java:
```diff
@@ -246,7 +257,12 @@ public Jvm getJvm(final String jvmName) {
-
+        List<WebServer> webServerList = webServerPersistenceService.getWebServers();
+        for (WebServer webserver : webServerList) {
+            if (updateJvmRequest.getNewJvmName().equals(webserver.getName())) {
```
Use equalsIgnoreCase since we don't want the names to be case sensitive

jeddcuison commented on jwala-services/src/main/java/com/cerner/jwala/service/jvm/impl/JvmServiceImpl.java:
```diff
@@ -622,7 +638,15 @@ private void secureCopyJvmConfigJar(Jvm jvm, String jvmConfigJar, User user) thr
 
     private void deployJvmConfigJar(Jvm jvm, User user, String jvmJar) throws CommandFailureException {
```
My bad I missed some of the code reviews. I don't think this should be in a state service, please ask @akinny if we can transfer this

jeddcuison commented on jwala-services/src/main/java/com/cerner/jwala/service/jvm/impl/JvmServiceImpl.java:
```diff
@@ -622,7 +638,15 @@ private void secureCopyJvmConfigJar(Jvm jvm, String jvmConfigJar, User user) thr
     private void deployJvmConfigJar(Jvm jvm, User user, String jvmJar) throws CommandFailureException {
         final String parentDir = ApplicationProperties.get("remote.paths.instances");
-        CommandOutput execData = jvmControlService.controlJvm(
+        CommandOutput execData = jvmControlService.executeCreateDirectoryCommand(jvm, parentDir+"/"+jvm.getJvmName());
```
Be consistent with spacing

jmahmoodcc replied:
Was this code block added as part of a merge?
I don't see how this pertains to solving the problem, and I consider it to be a risky change at this point in the release.

akinny replied:
I deleted the following as part of my fix, can you double check with master? This will cause the creation of additional empty jvm  backup folder, I think you need to remove this.
jvmControlService.executeCreateDirectoryCommand(jvm, parentDir+"/"+jvm.getJvmName());

jeddcuison commented on jwala-services/src/main/java/com/cerner/jwala/service/jvm/impl/JvmServiceImpl.java:
```diff
@@ -622,7 +638,15 @@ private void secureCopyJvmConfigJar(Jvm jvm, String jvmConfigJar, User user) thr
+        } else {
+            String standardError = execData.getStandardError().isEmpty() ? execData.getStandardOutput() : execData.getStandardError();
+            LOGGER.error("Deploy command completed with error trying to extract and back up JVM config {} :: ERROR: {}", jvm.getJvmName(), standardError);
+            throw new InternalErrorException(FaultType.REMOTE_COMMAND_FAILURE, standardError.isEmpty() ? CommandOutputReturnCode.fromReturnCode(execData.getReturnCode().getReturnCode()).getDesc() : standardError);
```
I know InternalErrorException has been used in this application for some time now, I'm not in favor of it, my 2 cents is to throw an service exception that relates to this class. The reason is when you see the exception, you would immediately see InternalErrorException - it does not tell you right away where the problem happened.

jeddcuison commented on jwala-services/src/main/java/com/cerner/jwala/service/webserver/impl/WebServerServiceImpl.java:
```diff
@@ -50,17 +53,21 @@
                                 final InMemoryStateManagerService<Identifier<WebServer>, WebServerReachableState> inMemoryStateManagerService,
                                 final String templatePath,
-                                final BinaryDistributionLockManager binaryDistributionLockManager) {
+                                final BinaryDistributionLockManager binaryDistributionLockManager,
```
Ask the team if you should use class variable auto-wiring since that is what we talked about doing going forward

jmahmoodcc replied:
I would prefer all new construction injection parameters be auto wired. But like Jedd said earlier, this might require a lot of re-work in the unit test.

jeddcuison commented on jwala-services/src/main/java/com/cerner/jwala/service/webserver/impl/WebServerServiceImpl.java:
```diff
@@ -76,6 +83,13 @@ public WebServer createWebServer(final CreateWebServerRequest createWebServerReq
 
+        List<Jvm> jvmList = jvmPersistenceService.getJvms();
+        for (Jvm jvm : jvmList) {
+            if(createWebServerRequest.getName().equals(jvm.getJvmName())){
```
Use equalsIgnoreCase

jeddcuison commented on jwala-services/src/main/java/com/cerner/jwala/service/webserver/impl/WebServerServiceImpl.java:
```diff
@@ -134,6 +148,12 @@ public WebServer updateWebServer(final UpdateWebServerRequest anUpdateWebServerC
 
+        List<Jvm> jvmList = jvmPersistenceService.getJvms();
+        for (Jvm jvm : jvmList) {
+            if(anUpdateWebServerCommand.getNewName().equals(jvm.getJvmName())){
```
Use equalsIgnoreCase

jeddcuison commented on jwala-services/src/main/java/com/cerner/jwala/service/jvm/impl/JvmServiceImpl.java:
```diff
@@ -246,7 +257,12 @@ public Jvm getJvm(final String jvmName) {
+        List<WebServer> webServerList = webServerPersistenceService.getWebServers();
+        for (WebServer webserver : webServerList) {
+            if (updateJvmRequest.getNewJvmName().equals(webserver.getName())) {
+                throw new EntityExistsException("Webserver already exists with this name "+ updateJvmRequest.getNewJvmName());
```
Throw JvmServiceException

jeddcuison reviewed (changes requested):
- Use equalsIgnoreCase since names should not be case sensitive
- Write unit tests for these fixes

jeddcuison committed:
Make delete web server in configurations tab delete only new web servers

jeddcuison committed:
Make delete JVM in configurations tab delete only new JVMs

jeddcuison committed:
Return status 204 since we're not returning any content after web server deletion

rahulsayini committed:
add logger statements

PeterTeunissen commented:
Just my $0.02: I don't understand why people format their code manually.... Control-Shift-F is your friend.... (oh... and by the way: so is Control-Shift-O). You can hook it all up to Control-S while you save the file....

rahulsayini committed:
improve indentation and replace to equalIgnoreCase

PeterTeunissen commented on jwala-services/src/main/java/com/cerner/jwala/service/webserver/impl/WebServerServiceImpl.java:
```diff
@@ -76,6 +83,14 @@ public WebServer createWebServer(final CreateWebServerRequest createWebServerReq
         createWebServerRequest.validate();
 
+        List<Jvm> jvmList = jvmPersistenceService.getJvms();
+        for (Jvm jvm : jvmList) {
```
Same here: why not call findJvmByExactName method here?

PeterTeunissen reviewed (changes requested):
See comments: instead of getting lists and looping, why not try to get by name?

rahulsayini committed:
use findJvmByExactName and findJvmByExactName instead of get all jvms and webservers

PeterTeunissen commented on jwala-services/src/test/java/com/cerner/jwala/service/jvm/impl/JvmServiceImplTest.java:
```diff
@@ -1006,7 +1009,7 @@ public JvmService getJvmService() {
+                    mockHistoryFacadeService, mockWebServerPersistenceService, new FileUtility());
         }
 
     }
```
You made this class compile, but where is the junit test that actually tests that trying to add a dupe will throw an error?

PeterTeunissen commented on jwala-services/src/test/java/com/cerner/jwala/service/webserver/impl/WebServerServiceImplTest.java:
```diff
@@ -132,7 +136,7 @@ public void setUp() throws IOException {
         mockWebServers12.add(mockWebServer2);
 
-        wsService = new WebServerServiceImpl(webServerPersistenceService, resourceService, inMemService, StringUtils.EMPTY, binaryDistributionLockManager);
+        wsService = new WebServerServiceImpl(webServerPersistenceService, resourceService, inMemService, StringUtils.EMPTY, binaryDistributionLockManager, jvmPersistenceService);
```
You made this class compile, but where is the junit test that actually tests that trying to add a dupe will throw an error?

Sharvaribarve committed:
Overrid functions in certain cases for specific rules and wrote unit tests

Sharvaribarve committed:
Merge branch 'APISpecialCharacters' into RESTSpecialCharacters

rahulsayini committed:
add test cases for service name validation

jeddcuison committed:
Add delete service functionality to deleteJvm

In addition

- Add test cases for different delete scenarios

jmahmoodcc commented on jwala-services/src/main/java/com/cerner/jwala/service/webserver/impl/WebServerServiceImpl.java:
```diff
@@ -76,6 +83,13 @@ public WebServer createWebServer(final CreateWebServerRequest createWebServerReq
+        List<Jvm> jvmList = jvmPersistenceService.getJvms();
+        for (Jvm jvm : jvmList) {
+            if(createWebServerRequest.getName().equals(jvm.getJvmName())){
+                throw new EntityExistsException("Jvm already exists with this name "+ createWebServerRequest.getName());
```
Like Jedd stated earlier, throw a WebServerServiceException here. I'm guilty of this as well, but now I know what to look for in the future.

jmahmoodcc commented on jwala-services/src/main/java/com/cerner/jwala/service/jvm/impl/JvmServiceImpl.java:
```diff
@@ -126,13 +125,12 @@ public JvmServiceImpl(final JvmPersistenceService jvmPersistenceService,
+
+        if (null != webServerPersistenceService.findWebServerByName(aCreateJvmRequest.getJvmName())) {
+            LOGGER.error("Webserver already exists with this name {}", aCreateJvmRequest.getJvmName());
+            throw new EntityExistsException("Webserver already exists with this name "+ aCreateJvmRequest.getJvmName());
```
Change all the EntityExistsException to the name of the class exceptions

jmahmoodcc reviewed (changes requested):
I think you got most of the changes requested from Peter and Jedd. It looks like the only thing left are to throw different exceptions based on the class their in.

akinny reviewed (changes requested):
Make sure to check my comment

rahulsayini committed:
throw exceptions based on class

jeddcuison committed:
Add delete service functionality to deleteWebServer

In addition

- Add test cases for different delete scenarios

jeddcuison committed:
Refactor

- Define variables common in several methods as class variables
- Remove unneeded setting and clearing of application properties root path

jeddcuison committed:
Add delete web server and delete JVM functionality in the group operations page

jeddcuison committed:
Add $1 to query only the service being deleted

jeddcuison committed:
Change log error to debug

Sharvaribarve committed:
Addressed code review by taking care of leading/trailing white spaces and added cases for backslash

rahulsayini committed:
use autowired instead of constructor

rahulsayini committed:
check for existing associations before remove of media

jeddcuison committed:
Change query parameter and add test for 204

rahulsayini committed:
add a logger and edit error message

rahulsayini committed:
add test case for media delete dependency

jeddcuison committed:
Log hardDelete and make it final

jeddcuison committed:
Change logging level to warn and error and remove duplicate comment

rahulsayini committed:
change equals to equalsIgnoreCase

PeterTeunissen committed:
Fixes for the hct.xml at the webapp level.

PeterTeunissen committed:
Merge pull request #48 from cerner/fix_webapp_resource_deploy

Fixes for the hct.xml at the webapp level.

Sharvaribarve committed:
Changed 12 hour to 24 hour format

LinHungWu committed:
updating buildVersion for JWALA to 0.0.185 for release

LinHungWu committed:
updating buildVersion to next snapshot version 0.0.186-SNAPSHOT

LinHungWu committed:
Uploading non-war file in web archive of web app instead of War file------>No error thrown + file uploaded on remote server.

jeddcuison committed:
Remove "tr<space>" to make element querying faster

jeddcuison committed:
Merge pull request #42 from cerner/make-jvm-webserver-delete-consistent

Make jvm webserver delete consistent

rahulsayini committed:
change to use promisedDel and address review comments

akinny committed:
Merge pull request #51 from cerner/SortedDates

Changed 12 hour to 24 hour format

akinny committed:
Merge pull request #49 from cerner/upload-resource-war

Uploading non-war file in web archive of web app instead of War file

Sharvaribarve committed:
Changed comment format

Sharvaribarve committed:
Reformatted comments

rahulsayini committed:
revert back the changes and update MediaConfig.js confirmDeleteCallback

akinny committed:
Merge pull request #44 from cerner/delete-media-throw-dependency-error

check for existing associations before removing media

jmahmoodcc committed:
Merge pull request #39 from cerner/RESTSpecialCharacters

Incorporate backend validation of names for groups,web-servers and JVMs

rahulsayini committed:
add test cases for WebServerServiceImpl.java

jmahmoodcc committed:
remove unused parameter when constructing WebServerServiceImpl

rahulsayini committed:
Merge branch 'master' into cannot-have-same-jvm-webserver-service-name

# Conflicts:
#	jwala-services/src/main/java/com/cerner/jwala/service/jvm/impl/JvmServiceImpl.java
#	jwala-services/src/test/java/com/cerner/jwala/service/jvm/impl/JvmServiceImplTest.java
#	jwala-services/src/test/java/com/cerner/jwala/service/webserver/impl/WebServerServiceImplTest.java

rahulsayini committed:
Merge branch 'cannot-have-same-jvm-webserver-service-name' of https://github.com/cerner/jwala into cannot-have-same-jvm-webserver-service-name
